lua/obsidian/cache.lua

```diff
@@ -0,0 +1,305 @@
+local async = require "plenary.async"
+local abc = require "obsidian.abc"
+local search = require "obsidian.search"
+local Note = require "obsidian.note"
+local log = require "obsidian.log"
+local EventTypes = require("obsidian.filewatch").EventTypes
+local uv = vim.uv
+local os_util = require("obsidian.os_util")
+
+---This class allows you to find the notes in your vault more quickly.
+---It scans your vault and saves the founded metadata to the file specified in your CacheOpts (by default it's ".cache.json").
+---For now, this allows to search for alises of your
+---notes in much shorter time.
+---@class obsidian.Cache : obsidian.ABC
+---
+---@field client obsidian.Client
+local Cache = abc.new_class()
+
+---Contains some information from the metadata of your note plus additional info.
+---@class obsidian.cache.CacheNote
+---
+---@field absolute_path string The full path to the note.
+---@field relative_path string The relative path to the root of the vault.
+---@field aliases string[] The alises of the note founded in the frontmatter.
+---@field last_updated number The last time the note was updated in seconds since epoch.
+
+---Converts the cache to JSON and saves to the file at the given path.
+---@param cache_notes { [string]: obsidian.cache.CacheNote } Dictionary where key is the relative path and value is the cache of the note.
+---@param cache_file_path string|obsidian.Note Location to save the cache
+local save_cache_notes_to_file = function(cache_notes, cache_file_path)
+  local save_path
+
+  if type(cache_file_path) == "obsidian.Note" then
+    save_path = cache_file_path.path.filename
+  else
+    save_path = cache_file_path
+  end
+
+  local file, err = io.open(save_path, "w")
+
+  if file then
+    file:write(vim.fn.json_encode(cache_notes))
+    file:close()
+  else
+    error(table.concat { "Couldn't write vault index to the file: ", save_path, ". Description: ", err })
+  end
+end
+
+---Update the cache file for changed files.
+---@param self obsidian.Cache
+---@return fun (changed_files: obsidian.filewatch.CallbackArgs[])
+local create_on_file_change_callback = function(self)
+  return function(changed_files)
+    vim.schedule(function()
+      local ok, cache_notes = pcall(self.get_cache_notes_from_file, self)
```

Comment:
no need to use pcall, since the function would not throw an error, just do something like, since we always want the cache file to exist at this point.
```suggestion
      local cache_notes = assert(self:get_cache_notes_from_file(), "An error occured when reading from the cache file.")
```